Part2

mastrHyperion98 opened the pull request:
Part 2 is completed minus the changes from Part4 (game end)

Here is a sample output: Please test the this vigorously. Part 2 is big and adds a lot of significant changes. 
```

"/home/hyperion/Documents/University/Winter_2020/Comp 345_C++/Project/Comp345Project/NewHaven/cmake-build-debug/Part2"
Enter the number of players (2, 3 or 4): 2
LOADING ../config/GBMapConfig_0.config
LOADING SUCCESSFUL
CREATING 2 PLAYERS!
Enter your 8-digit student ID number: 40057065
Enter your 8-digit student ID number: 12345678
2 PLAYERS HAVE BEEN SUCCESSFULLY CREATED!
LOADING ../config/VGMapNum_0.config
LOADING SUCCESSFUL
LOADING ../config/VGMapNum_1.config
LOADING SUCCESSFUL
SETTING GAME RESOURCE MARKERS!
GAME RESOURCE MARKERS SET SUCCESSFULLY
CREATING BUILDING DECK!
BUILDING DECK CREATED SUCCESSFULLY!
CREATING HARVEST DECK!
HARVEST DECK CREATED SUCCESSFULLY!
NEW PLAYER GETTING READY TO DRAW HIS/HER INTIAL SETUP!
DRAWING A BUILDING
DRAWING A BUILDING
DRAWING A BUILDING
DRAWING A BUILDING
DRAWING A BUILDING
DRAWING A HARVEST TILE
DRAWING A HARVEST TILE
DRAWING A HARVEST TILE
THE NEW PLAYER HAS FINISHED DRAWING THEIR HARVEST TILES AND BUILDINGS!
NEW PLAYER GETTING READY TO DRAW HIS/HER INTIAL SETUP!
DRAWING A BUILDING
DRAWING A BUILDING
DRAWING A BUILDING
DRAWING A BUILDING
DRAWING A BUILDING
DRAWING A HARVEST TILE
DRAWING A HARVEST TILE
DRAWING A HARVEST TILE
THE NEW PLAYER HAS FINISHED DRAWING THEIR HARVEST TILES AND BUILDINGS!
INITIALIZATION SUCCESSFUL!

























STARTING PLAYER HAS ID: 12345678
***GAME BOARD ID CONFIGURATION***
 -  00 01 02 03 04 -
 -  05 06 07 08 09 -
 -  10 11 12 13 14 -
 -  15 16 17 18 19 -
 -  20 21 22 23 24 -

***GAME BOARD CONTENT***
ST  SH    01  01    02  02    03  03    WT  SH    
WD  WD    01  01    02  02    03  03    WD  WT    

05  05    06  06    07  07    08  08    09  09    
05  05    06  06    07  07    08  08    09  09    

10  10    11  11    12  12    13  13    14  14    
10  10    11  11    12  12    13  13    14  14    

15  15    16  16    17  17    18  18    19  19    
15  15    16  16    17  17    18  18    19  19    

ST  ST    21  21    22  22    23  23    SH  ST    
WT  WD    21  21    22  22    23  23    WT  SH    


12345678 Your turn! What would you like to do? Enter the number for the move you'd like to make.
1	Play a Harvest tile from your possession
2	Play with your Shipment tile.
Choice: 1
tile index: 0 content	Sheep , Stone , Wood , Sheep
tile index: 1 content	Wood , Wheat , Wood , Sheep
Tile index to place: 1
position index to place tile: 9
***UPDATED GAME BOARD CONTENT***
ST  SH    01  01    02  02    03  03    WT  SH    
WD  WD    01  01    02  02    03  03    WD  WT    

05  05    06  06    07  07    08  08    WD  WT    
05  05    06  06    07  07    08  08    SH  WD    

10  10    11  11    12  12    13  13    14  14    
10  10    11  11    12  12    13  13    14  14    

15  15    16  16    17  17    18  18    19  19    
15  15    16  16    17  17    18  18    19  19    

ST  ST    21  21    22  22    23  23    SH  ST    
WT  WD    21  21    22  22    23  23    WT  SH    

                    ***STRATFORD***        
-------------------------------------------------------------------------
        6 |    00    01    02    03    04         | #Colonists: 06
        5 |    05    06    07    08    09         | #Colonists: 05
        4 |    10    11    12    13    14         | #Colonists: 04
        3 |    15    16    17    18    19         | #Colonists: 03
        2 |    20    21    22    23    24         | #Colonists: 02
        1 |    25    26    27    28    29         | #Colonists: 01
-------------------------------------------------------------------------
#Colonists|    05    04    03    04    05    

Available resources:

Sheep:	1

Stone:	0

Wheat:	2

Wood:	3
building index: 1 type	Wheat cost: 3
building index: 2 type	Wood cost: 4
building index: 3 type	Wheat cost: 4
building index: 4 type	Sheep cost: 3
building index: 5 type	Wheat cost: 1
Building index to play (0 to skip): 5
position index to place tile: 29
Do you want to place the building face up or face down? (0 for face up or 1 for face down): 
0
Select one of the options below: 
1	Play another building from your possession
2	Done building.
Choice: 1
                    ***STRATFORD***        
-------------------------------------------------------------------------
        6 |    00    01    02    03    04         | #Colonists: 06
        5 |    05    06    07    08    09         | #Colonists: 05
        4 |    10    11    12    13    14         | #Colonists: 04
        3 |    15    16    17    18    19         | #Colonists: 03
        2 |    20    21    22    23    24         | #Colonists: 02
        1 |    25    26    27    28    WT         | #Colonists: 01
-------------------------------------------------------------------------
#Colonists|    05    04    03    04    05    

Available resources:

Sheep:	1

Stone:	0

Wheat:	1

Wood:	3
building index: 1 type	Wheat cost: 3
building index: 2 type	Wood cost: 4
building index: 3 type	Wheat cost: 4
building index: 4 type	Sheep cost: 3
Building index to play (0 to skip): 0
40057065 Would you like to use the leftover resources to erect a building in your village?
1	Erect a building.
2	Pass
Choice: 1
                    ***GUILFORD***        
-------------------------------------------------------------------------
        6 |    00    01    02    03    04         | #Colonists: 06
        5 |    05    06    07    08    09         | #Colonists: 05
        4 |    10    11    12    13    14         | #Colonists: 04
        3 |    15    16    17    18    19         | #Colonists: 03
        2 |    20    21    22    23    24         | #Colonists: 02
        1 |    25    26    27    28    29         | #Colonists: 01
-------------------------------------------------------------------------
#Colonists|    05    04    03    04    05    

Available resources:

Sheep:	1

Stone:	0

Wheat:	1

Wood:	3
building index: 1 type	Stone cost: 1
building index: 2 type	Wood cost: 3
building index: 3 type	Wheat cost: 6
building index: 4 type	Wheat cost: 3
building index: 5 type	Sheep cost: 3
Building index to play (0 to skip): 2
position index to place tile: 16
Do you want to place the building face up or face down? (0 for face up or 1 for face down): 
0
Select one of the options below: 
1	Play another building from your possession
2	Done building.
Choice: 1
                    ***GUILFORD***        
-------------------------------------------------------------------------
        6 |    00    01    02    03    04         | #Colonists: 06
        5 |    05    06    07    08    09         | #Colonists: 05
        4 |    10    11    12    13    14         | #Colonists: 04
        3 |    15    WD    17    18    19         | #Colonists: 03
        2 |    20    21    22    23    24         | #Colonists: 02
        1 |    25    26    27    28    29         | #Colonists: 01
-------------------------------------------------------------------------
#Colonists|    05    04    03    04    05    

Available resources:

Sheep:	1

Stone:	0

Wheat:	1

Wood:	0
building index: 1 type	Stone cost: 1
building index: 2 type	Wheat cost: 6
building index: 3 type	Wheat cost: 3
building index: 4 type	Sheep cost: 3
Building index to play (0 to skip): 0
Leftover resources:

Sheep:	1

Stone:	0

Wheat:	1

Wood:	0

***Building Pool Content***

building index: 1 type	Stone cost: 4
building index: 2 type	Wheat cost: 6
building index: 3 type	Stone cost: 1
building index: 4 type	Wheat cost: 2
building index: 5 type	Stone cost: 2

Enter building card index to draw from the pool: 2


Do you want to pick another card from the pool? (y/n): y


***Building Pool Content***

building index: 1 type	Stone cost: 4
building index: 2 type	Stone cost: 1
building index: 3 type	Wheat cost: 2
building index: 4 type	Stone cost: 2

Enter building card index to draw from the pool: 3

DRAWING A HARVEST TILE
***GAME BOARD ID CONFIGURATION***
 -  00 01 02 03 04 -
 -  05 06 07 08 09 -
 -  10 11 12 13 14 -
 -  15 16 17 18 19 -
 -  20 21 22 23 24 -

***GAME BOARD CONTENT***
ST  SH    01  01    02  02    03  03    WT  SH    
WD  WD    01  01    02  02    03  03    WD  WT    

05  05    06  06    07  07    08  08    WD  WT    
05  05    06  06    07  07    08  08    SH  WD    

10  10    11  11    12  12    13  13    14  14    
10  10    11  11    12  12    13  13    14  14    

15  15    16  16    17  17    18  18    19  19    
15  15    16  16    17  17    18  18    19  19    

ST  ST    21  21    22  22    23  23    SH  ST    
WT  WD    21  21    22  22    23  23    WT  SH    


40057065 Your turn! What would you like to do? Enter the number for the move you'd like to make.
1	Play a Harvest tile from your possession
2	Play with your Shipment tile.
Choice: 1
tile index: 0 content	Stone , Wheat , Stone , Wood
tile index: 1 content	Stone , Wheat , Wood , Stone
Tile index to place: 0
position index to place tile: 15
***UPDATED GAME BOARD CONTENT***
ST  SH    01  01    02  02    03  03    WT  SH    
WD  WD    01  01    02  02    03  03    WD  WT    

05  05    06  06    07  07    08  08    WD  WT    
05  05    06  06    07  07    08  08    SH  WD    

10  10    11  11    12  12    13  13    14  14    
10  10    11  11    12  12    13  13    14  14    

ST  WT    16  16    17  17    18  18    19  19    
WD  ST    16  16    17  17    18  18    19  19    

ST  ST    21  21    22  22    23  23    SH  ST    
WT  WD    21  21    22  22    23  23    WT  SH    

                    ***GUILFORD***        
-------------------------------------------------------------------------
        6 |    00    01    02    03    04         | #Colonists: 06
        5 |    05    06    07    08    09         | #Colonists: 05
        4 |    10    11    12    13    14         | #Colonists: 04
        3 |    15    WD    17    18    19         | #Colonists: 03
        2 |    20    21    22    23    24         | #Colonists: 02
        1 |    25    26    27    28    29         | #Colonists: 01
-------------------------------------------------------------------------
#Colonists|    05    04    03    04    05    

Available resources:

Sheep:	0

Stone:	4

Wheat:	1

Wood:	1
building index: 1 type	Stone cost: 1
building index: 2 type	Wheat cost: 6
building index: 3 type	Wheat cost: 3
building index: 4 type	Sheep cost: 3
Building index to play (0 to skip): 1
position index to place tile: 27
Do you want to place the building face up or face down? (0 for face up or 1 for face down): 
0
Select one of the options below: 
1	Play another building from your possession
2	Done building.
Choice: 1
                    ***GUILFORD***        
-------------------------------------------------------------------------
        6 |    00    01    02    03    04         | #Colonists: 06
        5 |    05    06    07    08    09         | #Colonists: 05
        4 |    10    11    12    13    14         | #Colonists: 04
        3 |    15    WD    17    18    19         | #Colonists: 03
        2 |    20    21    22    23    24         | #Colonists: 02
        1 |    25    26    ST    28    29         | #Colonists: 01
-------------------------------------------------------------------------
#Colonists|    05    04    03    04    05    

Available resources:

Sheep:	0

Stone:	3

Wheat:	1

Wood:	1
building index: 1 type	Wheat cost: 6
building index: 2 type	Wheat cost: 3
building index: 3 type	Sheep cost: 3
Building index to play (0 to skip): 0
12345678 Would you like to use the leftover resources to erect a building in your village?
1	Erect a building.
2	Pass
Choice: 1
                    ***STRATFORD***        
-------------------------------------------------------------------------
        6 |    00    01    02    03    04         | #Colonists: 06
        5 |    05    06    07    08    09         | #Colonists: 05
        4 |    10    11    12    13    14         | #Colonists: 04
        3 |    15    16    17    18    19         | #Colonists: 03
        2 |    20    21    22    23    24         | #Colonists: 02
        1 |    25    26    27    28    WT         | #Colonists: 01
-------------------------------------------------------------------------
#Colonists|    05    04    03    04    05    

Available resources:

Sheep:	0

Stone:	3

Wheat:	1

Wood:	1
building index: 1 type	Wheat cost: 3
building index: 2 type	Wood cost: 4
building index: 3 type	Wheat cost: 4
building index: 4 type	Sheep cost: 3
building index: 5 type	Wheat cost: 6
building index: 6 type	Wheat cost: 2
Building index to play (0 to skip): 0
Leftover resources:

Sheep:	0

Stone:	3

Wheat:	1

Wood:	1

***Building Pool Content***

building index: 1 type	Stone cost: 4
building index: 2 type	Stone cost: 1
building index: 3 type	Stone cost: 2
building index: 4 type	Wood cost: 1
building index: 5 type	Wheat cost: 5

Enter building card index to draw from the pool: 4

DRAWING A HARVEST TILE
***GAME BOARD ID CONFIGURATION***
 -  00 01 02 03 04 -
 -  05 06 07 08 09 -
 -  10 11 12 13 14 -
 -  15 16 17 18 19 -
 -  20 21 22 23 24 -

***GAME BOARD CONTENT***
ST  SH    01  01    02  02    03  03    WT  SH    
WD  WD    01  01    02  02    03  03    WD  WT    

05  05    06  06    07  07    08  08    WD  WT    
05  05    06  06    07  07    08  08    SH  WD    

10  10    11  11    12  12    13  13    14  14    
10  10    11  11    12  12    13  13    14  14    

ST  WT    16  16    17  17    18  18    19  19    
WD  ST    16  16    17  17    18  18    19  19    

ST  ST    21  21    22  22    23  23    SH  ST    
WT  WD    21  21    22  22    23  23    WT  SH    


12345678 Your turn! What would you like to do? Enter the number for the move you'd like to make.
1	Play a Harvest tile from your possession
2	Play with your Shipment tile.
Choice: 1
tile index: 0 content	Sheep , Stone , Wood , Sheep
tile index: 1 content	Wood , Wood , Stone , Wood
Tile index to place: 0
position index to place tile: 14
***UPDATED GAME BOARD CONTENT***
ST  SH    01  01    02  02    03  03    WT  SH    
WD  WD    01  01    02  02    03  03    WD  WT    

05  05    06  06    07  07    08  08    WD  WT    
05  05    06  06    07  07    08  08    SH  WD    

10  10    11  11    12  12    13  13    SH  ST    
10  10    11  11    12  12    13  13    SH  WD    

ST  WT    16  16    17  17    18  18    19  19    
WD  ST    16  16    17  17    18  18    19  19    

ST  ST    21  21    22  22    23  23    SH  ST    
WT  WD    21  21    22  22    23  23    WT  SH    

                    ***STRATFORD***        
-------------------------------------------------------------------------
        6 |    00    01    02    03    04         | #Colonists: 06
        5 |    05    06    07    08    09         | #Colonists: 05
        4 |    10    11    12    13    14         | #Colonists: 04
        3 |    15    16    17    18    19         | #Colonists: 03
        2 |    20    21    22    23    24         | #Colonists: 02
        1 |    25    26    27    28    WT         | #Colonists: 01
-------------------------------------------------------------------------
#Colonists|    05    04    03    04    05    

Available resources:

Sheep:	3

Stone:	1

Wheat:	0

Wood:	1
building index: 1 type	Wheat cost: 3
building index: 2 type	Wood cost: 4
building index: 3 type	Wheat cost: 4
building index: 4 type	Sheep cost: 3
building index: 5 type	Wheat cost: 6
building index: 6 type	Wheat cost: 2
Building index to play (0 to skip): 4
position index to place tile: 17
Do you want to place the building face up or face down? (0 for face up or 1 for face down): 
0
Select one of the options below: 
1	Play another building from your possession
2	Done building.
Choice: 2
40057065 Would you like to use the leftover resources to erect a building in your village?
1	Erect a building.
2	Pass
Choice: 1 
```

mastrHyperion98 committed:
Merge branch 'Assignment2' into Part2

mastrHyperion98 committed:
Forgot to set tiles as having been played

mastrHyperion98 committed:
Create barebone GameController

mastrHyperion98 committed:
Create UnititializedControllerException

mastrHyperion98 committed:
Barebone cpp implementation of GameController

mastrHyperion98 committed:
Update

mastrHyperion98 committed:
Update player and GBMap(shouldnt be singleton)

mastrHyperion98 committed:
More updates and refactoring

mastrHyperion98 committed:
Determine starting player

mastrHyperion98 committed:
Make HarvestTile resourceTypes array public

mastrHyperion98 committed:
Update / nattery is dying on laptop so switching to PC 

yup one of those useless comiits

mastrHyperion98 committed:
Major changes/refactoring

mastrHyperion98 committed:
Update prior to merge

mastrHyperion98 committed:
Merge branch 'Assignment2' into Part2

mastrHyperion98 committed:
Performed some changes that might break existing drivers slightly

mastrHyperion98 committed:
Fixed bugs preventing Part1 Driver from running due to changes

mastrHyperion98 committed:
Update CMake catch by reference and Part2Driver

mastrHyperion98 committed:
Print out starting player

mastrHyperion98 committed:
Update Controller, add casting functions and printing for Game Board

mastrHyperion98 committed:
Updates

mastrHyperion98 committed:
Fix build Village input bug

mastrHyperion98 committed:
Add header guard to player class

mastrHyperion98 committed:
Completed verification/testing of components. Everything works! J

mastrHyperion98 committed:
Done Part2

MahChaa commented on NewHaven/src/GBMap.cpp:
```diff
@@ -115,7 +110,132 @@ ResourceTrails * GBMap::getResourcedGraph(int position) {
+   */
+  string spacer{"    "};
+  string inner_spacer("  ");
+    // Check config to determine #row
```
These should be made constants.

MahChaa commented on NewHaven/src/GBMap.cpp:
```diff
@@ -115,7 +110,132 @@ ResourceTrails * GBMap::getResourcedGraph(int position) {
+  string inner_spacer("  ");
+    // Check config to determine #row
+    int row_num{5};
+    int row_per_tile{2};
```
`row_per_tile` is never used.

MahChaa commented on NewHaven/src/GBMap.cpp:
```diff
@@ -115,7 +110,132 @@ ResourceTrails * GBMap::getResourcedGraph(int position) {
+        row_num = 7;
+    // we will iterate by row
+    if(*CONFIG != 2){
+        int num_per_row{5};
```
`num_per_row` is a bit confusing. I believe you meant `col_num`.

MahChaa commented on NewHaven/src/GBMap.cpp:
```diff
@@ -115,7 +110,132 @@ ResourceTrails * GBMap::getResourcedGraph(int position) {
+    if(*CONFIG != 2){
+        int num_per_row{5};
+        for(int i{0}; i < row_num; i++){
+            for(int j{i*5}; j <  i*num_per_row + num_per_row; j++){
```
Use parentheses to avoid operator precedence confusion. Better readability.

MahChaa commented on NewHaven/src/GBMap.cpp:
```diff
@@ -263,36 +383,67 @@ void GBMap::assignDefaultTiles() {
             placeHarvestTile(0, tl);
+            *(*board)[0].isPlayed = true;
             placeHarvestTile(4,tr);
-            placeHarvestTile(20, bl);
-            placeHarvestTile(24, br);
+            *(*board)[4].isPlayed = true;
+            placeHarvestTile(20, br);
+            *(*board)[20].isPlayed = true;
+            placeHarvestTile(24, bl);
+            *(*board)[24].isPlayed = true;
```
Good catch. This will avoid us some problems with tile placement.

MahChaa committed:
Performed the required changes from the PR to GBMap.cpp

MahChaa commented on NewHaven/src/GameController.cpp:
```diff
@@ -0,0 +1,262 @@
+current_turn_player{new int(*controller.current_turn_player)},  game_settings{nullptr}{
+    delete current;
+    current = this;
+}
```
You mean copy constructor. Why don't we want to copy `game_settings` too? Also, we shouldn't allow copying.

MahChaa commented on NewHaven/src/GameController.cpp:
```diff
@@ -0,0 +1,262 @@
+        *game_settings = *controller.game_settings;
+
+    return *this;
+}
```
This will never happen, you need two instances to use an assignment operator, which we can't have in this case. If we do, we will get an error.

MahChaa committed:
Created a function for the Hand class to print the harvest tiles in hand.

MahChaa committed:
Fixed small index mistake for the printing of harvest tiles in hand.

MahChaa committed:
Made use of the harvestTiles printing function in Hand in placeHarvestTile in Player. Also adjusted indexes to start at 1 instead of 0.

moohere reviewed (dismissed):
Everything legit looks solid, didn't get any errors, once we put the entire assignment2 together, will retest.

MahChaa commented on NewHaven/src/GameController.cpp:
```diff
@@ -0,0 +1,262 @@
+    game_settings->board->printBoard();
+    cout << endl;
+    cout << (*game_settings->players)[*current_turn_player]->getID() << " Your turn! What would you like to do? "
+            "Enter the number for the move you'd like to make." << endl;
```
We shouldn't spam the player with two options all the time even after he already played his shipment tile. It get annoying to have only one valid choice in an option menu.

MahChaa replied:
The player should also see his cards before he has to make a decision.

MahChaa committed:
Small formatting fix to printHarvestTiles() in Hand.

MahChaa commented on NewHaven/src/GameController.cpp:
```diff
@@ -0,0 +1,262 @@
+
+        cout << "Leftover resources:" << endl;
+        game_settings->tracker->printScore();
+
```
The player should be shown the leftover resources before he is given the option to erect buildings, not after.

MahChaa commented on NewHaven/src/Resources.h:
```diff
@@ -10,10 +10,8 @@ std::ostream& operator<<(std::ostream& output, const ResourceTypes& resource); /
-
 public:
-
+    ResourceTypes* tileContent;	//Array of resource types
 	HarvestTile(ResourceTypes*);
     HarvestTile(const HarvestTile &tile);
 	~HarvestTile();
```
So `tileContent` is public now? We won't need `getTitleContent()` anymore?

MahChaa commented on NewHaven/src/Setting.cpp:
```diff
@@ -9,22 +9,18 @@
 Setting::Setting(const Setting& setting):h_deck{new HarvestDeck(*setting.h_deck)}, b_deck{new BuildingDeck(*setting.b_deck)},
-    board{new GBMap(*setting.board)}, players{new vector<Player>(*setting.players)}{
+    board{new GBMap(*setting.board)}, players{new vector<Player*>(*setting.players)}, tracker{new ResourceTracker{*setting.tracker}
+}{
     // singleton design we dont need any other reference but the current one
-    delete current;
-    current = this;
 }
```
It isn't a singleton anymore. This copy constructor is very problematic. It doesn't make any `nullptr` checks. I won't break my head over the segmentation faults though, we'll never use it. It's not noticeable if you don't pay attention.

mastrHyperion98 replied:
Right I forgot to check that, either way we make a direct reference throught the GameController... we shouldnt ever make a copy but ill fix it nevertheless. We could possibly lose points

MahChaa commented on NewHaven/src/Setting.cpp:
```diff
@@ -9,22 +9,18 @@
-    current = this;
 }
 
 Setting& Setting::operator=(const Setting& setting){
```
The assignment operator is also broken, but we can overlook it for the same reasons the copy is constructor can be overlooked.

MahChaa commented on NewHaven/src/Setting.cpp:
```diff
@@ -214,14 +207,16 @@ bool Setting::initSetting() {
+            Player* it = (*players)[i];
             cout << "NEW PLAYER GETTING READY TO DRAW HIS/HER INTIAL SETUP!" << endl;
             for (int i = 0; i < 5; i++)
                 it->drawBuilding(drawBuilding());
```
lol that's lazy, just use `(*players)[i]->setVillage...` etc..

MahChaa commented on NewHaven/src/Setting.h:
```diff
@@ -17,10 +17,12 @@ class Setting {
+    vector<Player*>* players;
     bool initSetting();
+    BuildingDeck& getBuildingDeck();
+    void DrawBuilding(int);
```
`getBuildingDeck()` has no definition.

MahChaa commented on NewHaven/src/VGMap.cpp:
```diff
@@ -111,7 +113,44 @@ void VGMap::CreateVillageField() {
+    // content of row | row value
+    // column value|
+    string spacer = "    ";
+    string inner_spacer ="  ";
```
Constants

MahChaa commented on NewHaven/src/Player.cpp:
```diff
@@ -26,103 +26,121 @@ Player& Player::operator=(const Player &player) {
+                cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+                cout << "ERROR cannot play building in that position. Costs do not match!" << endl;
+                goto POSITION;
+            }
```
When you can't play a building because it's not adjacent to another building of the same type that was placed before and you don't have the resources necessary to place it beside a building of the same type, you are stuck in a loop from `POSITION` to here with no valid course of action and no way out.

MahChaa committed:
Made the PR required changes for Setting and some output formatting.

MahChaa committed:
Made PR required changes to VGMap.

MahChaa committed:
-Added a function to print the Buildings held in a hand to the hand class and to the player class.
-Made the PR required changes to Player and some output formatting fixes.

MahChaa reviewed (commented):
I made all the changes and fixes I mentioned in the PR and more. Confirm and I'll push the changes.

MahChaa committed:
-Deleted the copy contructor and the assignment operator overload of GameController.
-Modified the playTurn() function's formatting and behaviour to be more user friendly.
-Changed formatting and output printing order in ShareTheWealth().
-Made the PR requested changes.

MahChaa committed:
Small fix to isFlipped() of Building and tiny changes to Part2Driver.

MahChaa commented:
Also made a few formatting changes for better readability of outputs.

mastrHyperion98 committed:
Fix bug

mastrHyperion98 committed:
Fix copy constructor for Setting

mastrHyperion98 committed:
Add changes from part4

mastrHyperion98 committed:
Update the content from part 4

mastrHyperion98 committed:
Update output

mastrHyperion98 committed:
Update empty field counting

mastrHyperion98 committed:
Remove Broken and Old Drivers

mastrHyperion98 committed:
Fix End Game and Find Winner

mastrHyperion98 committed:
HarvestTile rotation

MahChaa commented on NewHaven/src/GameController.cpp:
```diff
@@ -137,6 +136,37 @@ void GameController::playTurn(){
+    announcement = "THE WINNER IS: ";
+
+else
+    announcement = "THE WINNERS ARE:\n";
```
You don't need to assign anything to the string `announcement` on declaration when you're not going to use the value assigned.

MahChaa commented on NewHaven/src/GameController.cpp:
```diff
@@ -254,4 +283,89 @@ void GameController::shareTheWealth(){
+        int score{(*it.base())->calculateScore()};
+        // add the entry to the map
+        player_scores.insert(entry(*it.base(),score));
+    }
```
`(*it.base())` doesn't compile. `it` doesn't have a `base()` member.

MahChaa commented on NewHaven/CMakeLists.txt:
```diff
@@ -14,11 +14,11 @@ if(UNIX)
-add_executable(GameBoardMapTest test/GBMapDriver.cpp src/GBMap.cpp src/Resources.cpp src/Square.cpp)
-add_executable(MapLoaderTest test/MapLoaderDriver.cpp src/VGMapLoader.cpp src/VGMap.cpp src/GBMapLoader.cpp src/GBMap.cpp src/Square.cpp Exceptions/InvalidConfigurationException.cpp Exceptions/BoardConfigurationNotLoaded.cpp src/Resources.cpp)
+add_executable(VGMap src/VGMap.cpp src/Resources.cpp)
+add_executable(GameBoardMapTest src/GBMap.cpp src/Resources.cpp src/Square.cpp)
+add_executable(MapLoaderTest src/VGMapLoader.cpp src/VGMap.cpp src/GBMapLoader.cpp src/GBMap.cpp src/Square.cpp Exceptions/InvalidConfigurationException.cpp Exceptions/BoardConfigurationNotLoaded.cpp src/Resources.cpp)
 add_executable(ResourcesTest test/ResourcesDriver.cpp src/Resources.cpp src/ResourceTracker.cpp)
-add_executable(ScoreTest test/ScoreDriver.cpp src/GBMap.cpp src/Resources.cpp src/Square.cpp src/VGMap.cpp src/ScoreCalculator.cpp src/ResourceTracker.cpp)
+add_executable(ScoreTest src/GBMap.cpp src/Resources.cpp src/Square.cpp src/VGMap.cpp src/ScoreCalculator.cpp src/ResourceTracker.cpp)
```
Why keep executables without `main()`? Weren't we going to remove all the driver that aren't relevant to asg2?

MahChaa commented on NewHaven/src/GBMap.cpp:
```diff
@@ -50,7 +50,7 @@ bool GBMap::placeHarvestTile(int NodeID, HarvestTile *tile) {
     *(*board)[NodeID].isPlayed = true;
-    *playCounter++;
+    *playCounter = *playCounter + 1;
     return true;
```
`*playCounter++` doesn't work because `++` doesn't mean the same thing on pointer than it does on integers. You'll need to dereference it before using `++` this way: `(*playCounter)++` or `++(*playCounter)`.

mastrHyperion98 replied:
As seen above this has already been resolved and I know i was incrementing the pointer address. It was an accident from lack of careful attention

MahChaa commented on NewHaven/src/GBMap.cpp:
```diff
@@ -437,5 +437,6 @@ HarvestTile * GBMap::getHarvestTile(int position) {
-    return (*SIZE - *playCounter) == 1;
+    const int LIMIT = 20;
+    return (*SIZE - *playCounter) == LIMIT;
 }
```
Why is this better than writing `20` directly into the bool evaluation?

MahChaa reviewed (changes requested):
Good implementation of the rotation mechanics. There's a compiling error that needs to be addressed. The copy and assignment operator of `Setting` cause seg faults (You should try to test them after you create/modify them) in the `ResourceTracker` assignment operator. I didn't look into it much. I haven't tested the Part2Driver much too. I'll do it when the fixes will be made. The copy constructor and assignment operator of `Setting` is really not a priority, they're never used.

mastrHyperion98 committed:
Fix Setting copy constructor

mastrHyperion98 committed:
More Setting Copy/Assignment Operator overload changes

mastrHyperion98 committed:
change it.base() -> it

mastrHyperion98 committed:
Remove other cases of it.base()

mastrHyperion98 committed:
it.base() -> it

mastrHyperion98 committed:
Change playCounter Limit to 1

mastrHyperion98 committed:
Tidy up a bit

mastrHyperion98 committed:
Remove all other non Assignment2 related drivers or no longer needed files

mastrHyperion98 committed:
Add comments

MahChaa commented on NewHaven/src/Setting.cpp:
```diff
@@ -56,15 +73,19 @@ Setting::~Setting() {
     for(int i = 0; i < numberOfPlayers;i++){
         string id{""};
-        while ((cout << "Enter your 8-digit student ID number: " && !(cin >> id))
+        while ((cout << "Enter your 8-digit student ID number(8 digits): " && !(cin >> id))
```
A bit redundant to write 8 digits twice in the instruction. I'll fix it.

MahChaa reviewed (commented):
All good.

mastrHyperion98 committed:
Revert change in prompt

mastrHyperion98 committed:
simplify and cleanup... remove unused code

mastrHyperion98 committed:
Print changes

MahChaa committed:
Small formatting fixes.

MahChaa committed:
Formatting fixes to Player.

mastrHyperion98 committed:
Push refactoring setBuilding

mastrHyperion98 committed:
Update formatting

MahChaa committed:
Fixed formatting with building pool printing.

MahChaa committed:
Merge branch 'Part2' of https://github.com/mastrHyperion98/Comp345Project into Part2
